### ISSUE TITLE
Masking for loss probably missing in loss_function.py

### ISSUE THREAD

**Issue by hitrefresh** [NVIDIA/tacotron2#80]
@rafaelvalle 
```
from torch import nn

class Tacotron2Loss(nn.Module):
    def __init__(self):
        super(Tacotron2Loss, self).__init__()

    def forward(self, model_output, targets):
        mel_target, gate_target = targets[0], targets[1]
        mel_target.requires_grad = False
        gate_target.requires_grad = False
        gate_target = gate_target.view(-1, 1)

        mel_out, mel_out_postnet, gate_out, _ = model_output
        gate_out = gate_out.view(-1, 1)
        mel_loss = nn.MSELoss()(mel_out, mel_target) + \
            nn.MSELoss()(mel_out_postnet, mel_target)
        gate_loss = nn.BCEWithLogitsLoss()(gate_out, gate_target)
        return mel_loss + gate_loss
```
https://github.com/NVIDIA/tacotron2/blob/master/loss_function.py

I checked in python debugger that `mel_out` and `mel_target` have size corresponding to the largest sequence. `mel_target` has padded zeros for shorter sequences while `mel_out` has 
values predicted after the sequence length for shorter sequences. The loss for shorter sequences
should be masked and not backpropagated.

**Comment by bfs18** [comment]
Hi, I am also curious about this. 
1. If the loss mask is provided, the gate label is extremely unbalanced. Only the last value of a sentence is 1 (end predicting) and the rest values in the same sentence are all 0's (continue predicting). So the end symbol would be very hard to predict if the loss mask is applied to the BCEWithLogitsLoss term.
2. If the loss mask is not provided, the padded 0's may be considered to align to the last symbol (may be a end of sentence symbol).
Besides, according to my experience, the model would learn alignment in much earlier iters if the loss mask is applied to the MSELoss term. The duration of the padding silence varies too much if the loss mask is not used, so the last symbol alignment is very hard to learn.

**Comment by hitrefresh** [comment]
Yes, that's a good point. In the current implementation, the loss mask is provided, but the duration of passing silence varies too much across sequences, because all the sequences in the batch are padded up to longest sequence. In theory, it can also happen that the same sequence can receive different padding based on the longest sequence in the minibatch sampled. 

I hope the contributors will be able to provide some thoughts on this. @rafaelvalle

**Comment by hitrefresh** [comment]
From Tacotron 1 paper:
`We train using a batch size of 32, where all sequences are padded to a max length. It’s a common
practice to train sequence models with a loss mask, which masks loss on zero-padded frames.
However, we found that models trained this way don’t know when to stop emitting outputs, causing
repeated sounds towards the end. One simple trick to get around this problem is to also reconstruct
the zero-padded frames.`
